### PR TITLE
Fixes for Julia 0.7 (and drop support for a range of DEV versions)

### PR DESCRIPTION
**Pull request by martinholters**
This can be cleanup a little bit  once JuliaLang/Compat.jl#463 is merged and tagged, which hopefully shouldn't take too long.

### CODE REVIEW

File: src/definitions.jl [hunk 1]
```diff
@@ -33,11 +41,11 @@ realfloat(x::AbstractArray{T}) where {T<:Real} = copy1(typeof(fftfloat(zero(T)))
 
 # copy to a 1-based array, using circular permutation
 function copy1(::Type{T}, x) where T
-    y = Array{T}(map(length, indices(x)))
+    y = Array{T}(uninitialized, map(length, axes(x)))
```
Comment (ararslan) [inline src/definitions.jl]: I think this will need to be `Compat.axes` for 0.6 compatibility. (It had to be unexported from Compat to avoid breaking AxisArrays et al.)

Comment (martinholters) [reply src/definitions.jl]: Ah, yes, of course ... except that it passed CI, although the used Compat 0.46 should not be exporting it. 😕 
Will change anyway.

**Comment by martinholters** [comment]
Good to go from my side.

File: src/definitions.jl [hunk 2]
```diff
@@ -1,8 +1,11 @@
 # This file was formerly a part of Julia. License is MIT: https://julialang.org/license
 
-using Base.LinAlg: BlasReal
+using Compat
+using Compat.LinearAlgebra
+using Compat.LinearAlgebra: BlasReal
+import Compat.LinearAlgebra: A_mul_B!, A_ldiv_B!
```
Comment (ararslan) [inline src/definitions.jl]: These are deprecated in 0.7. We'll have to redesign the API to use `mul!`.

Comment (martinholters) [reply src/definitions.jl]: Ah right, that change maybe was a bit too mechanical. But for compatibility with both Julia 0.6 and Julia 0.7, we need both `A_mul_B!`/`A_ldiv_B!` and `mul!`/`ldiv!`, right? I'll give it a sot.

Comment (appleparan) [reply src/definitions.jl]: Julia 0.6 use `Base.fft`

Comment (ararslan) [reply src/definitions.jl]: Since FFTW doesn't support 0.6 it might be okay to just move to 0.7 only here as well. You could put the Julia lower bound at 0.7.0-DEV.3204, the commit that removed `A_mul_B` et al., then just do everything in terms of `mul!`.

Comment (martinholters) [reply src/definitions.jl]: Ah yes, good point. Would also allow getting rid of all the Compat stuff. Yes, sounds good.

Comment (ararslan) [reply src/definitions.jl]: We'll still need Compat for a lot of the stdlib stuff though, since a lot of that came after the `A_mul_B` deprecation.

Comment (martinholters) [reply src/definitions.jl]: I took the easy way out and set the minimum version to 0.7.0-DEV.3449, that should indeed make Compat unnecessary.

**Comment by stevengj** [comment]
Seems weird to drop 0.6 support when 0.7 is not even released yet.

**Comment by martinholters** [comment]
True, but note that most of the code is behind `VERSION` conditionals and does almost nothing on 0.6 - only importing and re-exporting symbols. And the existing versions continue to do so perfectly on 0.6. Newer versions of `AbstractFFTs` are only needed for FFT implementations. Those would have to program against to slightly different APIs, depending on whether they run on 0.6 or 0.7. BUT: The only (?) existing implementation---FFTW.jl---is also effectively 0.7 only and has also bumped its requirement to 0.7.0-DEV.602. Or are there any other julia FFT implementations out there I'm not aware of?

What I could conceivably do is leave the 0.6 code in and only drop support for all 0.7 versions prior to 0.7.0-DEV.3449. However, I don't see the benefit, as 0.6 would get the same AbstractFFTs code as with previous versions, so why go through the hassle and make sure they can update?

**Comment by ararslan** [comment]
> Seems weird to drop 0.6 support when 0.7 is not even released yet.

Under the vast majority of circumstances (and for 99% of current packages) I would agree with that, but we're in an odd situation with this package, as Martin has laid out above, so I'm inclined to say it's okay here. I'll defer to your judgment though, @stevengj.

File: README.md [hunk 3]
```diff
@@ -25,9 +25,10 @@ To define a new FFT implementation in your own module, you should
 * Define a new method `AbstractFFTs.plan_fft(x, region; kws...)` that returns a `MyPlan` for at least some types of
   `x` and some set of dimensions `region`.
 
-* Define a method of `A_mul_B!(y, p::MyPlan, x)` that computes the transform `p` of `x` and stores the result in `y`.
+* Define a method of `LinearAlgebra.mul!(y, p::MyPlan, x)` that computes the transform `p` of `x` and stores the
```
Comment (martinholters) [inline README.md]: If we decide to not drop 0.6 support in this PR, should this instruct to add a method to `A_mul_B!(y, p::MyPlan, x)` or `LinearAlgebra.mul!(y, p::MyPlan, x)` depending on `VERSION`, or is there some uniform way to achieve that?

**Comment by martinholters** [comment]
Once we have come to a resolution and a new version has been tagged, we should put `0.7.0-DEV.3449` as a julia upper bound on all previous versions as this is the point where even `using` them starts to fail.

File: README.md (hunk 3 above)
Comment (ararslan) [reply README.md]: Since Compat doesn't have `mul!` in its LinearAlgebra submodule, I don't know that it's possible to do anything other than tell people to overload `A_mul_B!` on 0.6 and `mul!` on 0.7, unfortunately.

**Comment by martinholters** [comment]
TL;DR:
Proposal: Only drop support for `v"0.7.0-DEV.602" <= VERSION < v"0.7.0-DEV.3665"` (i.e. keep support for 0.6) and introduce `AbstractFFTs.execute!` (name subject to bikeshedding) for FFT implementations to extend independent of Julia version.

Here's my reasoning: Only if `VERSION >= v"0.7.0-DEV.602"`, anything non-trivial is being done by this package. However, lots of breaking changes have happened since then. We could either use `Compat` and `if VERSION` constructs throughout, or just stop supporting versions between `v"0.7.0-DEV.602"` and whatever 0.7.0-DEV version caused the latest relevant breaking change (`v"0.7.0-DEV.3665"` at present). I think I'd prefer the latter. Log term, I don't think anyone will keep using a -DEV version (and certainly shouldn't). Short-term, it would be nice to have a working AbstractFFTs for relatively recent master. Which we haven't had for some time. And with a sufficiently strict interpretation of "relatively recent", option 2 would be enough for that.

That leaves the question of what to do for `VERSION < v"0.7.0-DEV.602"` (especially 0.6). @stevengj didn't like dropping support (now), and I understand his point. The problem is the `A_mul_B!` to `mul!` rename, which can't be glossed over by some Compat magic. The alternative I do see is to provide an export (say `execute!`) in AbstractFFTs that is either an alias of `A_mul_B!` or `mul!` depending on Julia version. That way, any FFT implementations that target Julia 0.6 and later could just always extend `AbstractFFTs.execute!`. Downside: I don't think there are any FFT implementations out there that utilize AbstractFFTs and support Julia 0.6 (except with a hollow do-nothing in the case of FFTW). So we add this and make it part of the public interface although it might never actually be needed. (That's why I'd still prefer dropping 0.6 support.)

Thoughts?

**Comment by appleparan** [comment]
Totally agree. I'm looking forward to merge.

**Comment by martinholters** [comment]
I went with something intermediate: Only dropping support for a range of DEV versions, and mentioning the need to add a method to `A_mul_B!` instead of `mul!` on older Julia in the docs. Is that something we can agree on?

File: REQUIRE [hunk 4]
```diff
@@ -1 +1 @@
-julia 0.6
+julia 0.6 0.7.0-DEV.602 0.7.0-DEV.3449
```
Comment (ararslan) [inline REQUIRE]: This syntax is wrong; there should only be two versions listed. (How did this not cause an error?)

File: test/runtests.jl [hunk 5]
```diff
@@ -1,10 +1,15 @@
...
 using AbstractFFTs
-using Base.Test
-
-import AbstractFFTs: Plan, plan_fft, plan_inv, plan_bfft
-import Base: A_mul_B!, *
+using AbstractFFTs: Plan
+using Compat.LinearAlgebra
+using Compat.Test
+
+if VERSION < v"0.7.0-"
```
Comment (ararslan) [inline test/runtests.jl]: Could use a more precise version number

File: REQUIRE (hunk 4 above)
Comment (martinholters) [reply REQUIRE]: It's correct (but admittedly rarely used), see https://docs.julialang.org/en/latest/manual/packages/#Requirements-Specification-1. It means that this package may be installed if `v"0.6" <= VERSION < v"0.7.0-DEV.602" || v"0.7.0-DEV.3449" <= VERSION`.

File: src/definitions.jl (hunk 1 above)
Comment (martinholters) [reply src/definitions.jl]: To clarify: This file isn't even loaded on 0.6.

File: test/runtests.jl (hunk 5 above)
Comment (martinholters) [reply test/runtests.jl]: True, will fix.

File: src/definitions.jl [hunk 6]
```diff
@@ -245,7 +246,13 @@ size(p::ScaledPlan) = size(p.p)
 show(io::IO, p::ScaledPlan) = print(io, p.scale, " * ", p.p)
 summary(p::ScaledPlan) = string(p.scale, " * ", summary(p.p))
 
-*(p::ScaledPlan, x::AbstractArray) = scale!(p.p * x, p.scale)
+if VERSION >= v"0.7.0-DEV.3665"
+    *(p::ScaledPlan, x::AbstractArray) = LinearAlgebra.rmul!(p.p * x, p.scale)
```
Comment (martinholters) [inline src/definitions.jl]: Could also bump the (second) lower bound in REQUIRE to `0.7.0-DEV.3665` and do this unconditionally. Opinions?

File: REQUIRE (hunk 4 above)
Comment (ararslan) [reply REQUIRE]: Whoa, holy cow. TIL.

**Comment by martinholters** [comment]
Bump.

Is this OK as it is?

**Comment by ararslan** [comment]
Looks good to me if @stevengj is okay with it.

**Comment by martinholters** [comment]
Bumpity.
